File: server/makc2022--dotnet/Makc2022.Layer3.Sql.Sample/Entities/DummyTree/DummyTreeEntityLoader.cs
// Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

using Makc2022.Layer3.Sql.Sample.Entity;

namespace Makc2022.Layer3.Sql.Sample.Entities.DummyTree
{
    /// <summary>
    /// Загрузчик сущности "DummyTree".
    /// </summary>
    public class DummyTreeEntityLoader : EntityLoader<DummyTreeEntityObject>
    {
        #region Constructors

        /// <inheritdoc/>
        public DummyTreeEntityLoader(DummyTreeEntityObject? entityObject = null)
            : base(entityObject ?? new DummyTreeEntityObject())
        {
        }

        #endregion Constructors

        #region Public methods

        /// <inheritdoc/>
        public sealed override HashSet<string> Load(
            DummyTreeEntityObject entityObject,
            HashSet<string>? loadableProperties = null)
        {
            var result = base.Load(entityObject, loadableProperties);

            if (result.Contains(nameof(EntityObject.Id)))
            {
                EntityObject.Id = entityObject.Id;
            }

            if (result.Contains(nameof(EntityObject.Name)))
            {
                EntityObject.Name = entityObject.Name;
            }

            if (result.Contains(nameof(EntityObject.ParentId)))
            {
                EntityObject.ParentId = entityObject.ParentId;
            }

            if (result.Contains(nameof(EntityObject.TreeChildCount)))
            {
                EntityObject.TreeChildCount = entityObject.TreeChildCount;
            }

            if (result.Contains(nameof(EntityObject.TreeDescendantCount)))
            {
                EntityObject.TreeDescendantCount = entityObject.TreeDescendantCount;
            }

            if (result.Contains(nameof(EntityObject.TreeLevel)))
            {
                EntityObject.TreeLevel = entityObject.TreeLevel;
            }

            if (result.Contains(nameof(EntityObject.TreePath)))
            {
                EntityObject.TreePath = entityObject.TreePath;
            }

            if (result.Contains(nameof(EntityObject.TreePosition)))
            {
                EntityObject.TreePosition = entityObject.TreePosition;
            }

            if (result.Contains(nameof(EntityObject.TreeSort)))
            {
                EntityObject.TreeSort = entityObject.TreeSort;
            }

            return result;
        }

        #endregion Public methods

        #region Protected methods

        /// <inheritdoc/>
        protected override HashSet<string> CreateAllPropertiesToLoad()
        {
            return new HashSet<string>
            {
                nameof(EntityObject.Id),
                nameof(EntityObject.Name),
                nameof(EntityObject.ParentId),
                nameof(EntityObject.TreeChildCount),
                nameof(EntityObject.TreeDescendantCount),
                nameof(EntityObject.TreeLevel),
                nameof(EntityObject.TreePath),
                nameof(EntityObject.TreeSort)
            };
        }

        #endregion Protected methods
    }
}

File: server/makc2022--nodejs--nestjs/src/layer3/nosql-mongo/sample/entities/dummy-tree/dummy-tree-entity.loader.ts
/** Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License. */

import { nameof } from '../../../../../layer1/common/common-functions';
import { EntityLoader } from '../../../../../layer1/entity/entity-loader';
import { DummyTreeEntityObject } from './dummy-tree-entity.object';

/** Загрузчик сущности "DummyTree". */
export class DummyTreeEntityLoader extends EntityLoader<DummyTreeEntityObject> {
  /** @inheritdoc */
  constructor(entityObject?: DummyTreeEntityObject) {
    super(entityObject ?? ({} as DummyTreeEntityObject));
  }

  /** @inheritdoc */
  override load(
    entityObject: DummyTreeEntityObject,
    loadableProperties?: Set<string>
  ): Set<string> {
    const result = super.load(entityObject, loadableProperties);

    if (result.has(nameof<DummyTreeEntityObject>('id'))) {
      this.entityObject.id = entityObject.id;
    }

    if (result.has(nameof<DummyTreeEntityObject>('name'))) {
      this.entityObject.name = entityObject.name;
    }

    if (result.has(nameof<DummyTreeEntityObject>('parentId'))) {
      this.entityObject.parentId = entityObject.parentId;
    }

    if (result.has(nameof<DummyTreeEntityObject>('treeChildCount'))) {
      this.entityObject.treeChildCount = entityObject.treeChildCount;
    }

    if (result.has(nameof<DummyTreeEntityObject>('treeDescendantCount'))) {
      this.entityObject.treeDescendantCount = entityObject.treeDescendantCount;
    }

    if (result.has(nameof<DummyTreeEntityObject>('treeLevel'))) {
      this.entityObject.treeLevel = entityObject.treeLevel;
    }

    if (result.has(nameof<DummyTreeEntityObject>('treePath'))) {
      this.entityObject.treePath = entityObject.treePath;
    }

    if (result.has(nameof<DummyTreeEntityObject>('treePosition'))) {
      this.entityObject.treePosition = entityObject.treePosition;
    }

    if (result.has(nameof<DummyTreeEntityObject>('treeSort'))) {
      this.entityObject.treeSort = entityObject.treeSort;
    }

    return result;
  }

  /** @inheritdoc */
  protected override createAllPropertiesToLoad(): Set<string> {
    return new Set<string>([
      nameof<DummyTreeEntityObject>('id'),
      nameof<DummyTreeEntityObject>('name'),
      nameof<DummyTreeEntityObject>('parentId'),
      nameof<DummyTreeEntityObject>('treeChildCount'),
      nameof<DummyTreeEntityObject>('treeDescendantCount'),
      nameof<DummyTreeEntityObject>('treeLevel'),
      nameof<DummyTreeEntityObject>('treePath'),
      nameof<DummyTreeEntityObject>('treePosition'),
      nameof<DummyTreeEntityObject>('treeSort'),
    ]);
  }
}
